### ISSUE TITLE
install_asm fails to call meshconfig API

### ISSUE THREAD

**Issue by zhaytee** [GoogleCloudPlatform/anthos-service-mesh-packages#296]
Hello! I'm using install_asm as downloaded from the storage bucket here:

```
https://storage.googleapis.com/csm-artifacts/asm/install_asm_1.7
```

I've got a very simple setup. Just trying to install ASM into a single GKE cluster in one project.

I'm using a service account created specifically for this purpose.

The validation succeeds (i.e. running install_asm with --only_validate).

However, when I run the script to actually install ASM, I get a "401 Unauthorized" error when the script tries to access the meshconfig API. Here's how I'm invoking the script:

```
    ./install_asm -v \
          --project_id myprojectid \
          --cluster_name myclustername \
          --cluster_location us-central1-a \
          --mode install \
          --enable_apis \
          --service_account anthos-installer@myprojectid.iam.gserviceaccount.com \
          --key_file ./sa.json
```

It gets all the way down to "Initializing Mesh CA", and then fails with a "401 Unauthorized" as below:

```
install_asm: Initializing Mesh CA...
install_asm: Running: 'curl --request POST --fail --data  -o /dev/null https://meshconfig.googleapis.com/v1alpha1/projects/myprojectid:initialize --header @-'
install_asm: -------------
curl: (22) The requested URL returned error: 401 Unauthorized
```

I notice that the script is granting the following roles to the service account:

```
editor
compute.admin
container.admin
resourcemanager.projectIamAdmin
iam.serviceAccountAdmin
iam.serviceAccountKeyAdmin
gkehub.admin
```

However, the script does not grant `meshconfig.admin`. I'm guessing that that might be required here?

Also interesting: I see no roles related to the `meshconfig` API in the Google Cloud web console for my project, not even when I search for the `meshconfig` string in the Roles section.

The `meshconfig` API is fully enabled for my project.

I'm at a bit of a loss here. Any help would be appreciated. Thank you!

**Comment by zerobfd** [comment]
It shouldn't be `meshconfig.admin`, since I believe that's included in the `editor` role. (The role isn't listed in the web console for some technical reasons that aren't important, that's expected)

This might have been the fix that recently got rolled into the branch, but I was waiting on a particular customer to OK it for their test case before changing the version in the bucket. Could you try the version from this link and see if that works?
https://raw.githubusercontent.com/GoogleCloudPlatform/anthos-service-mesh-packages/release-1.7-asm/scripts/asm-installer/install_asm

**Comment by zhaytee** [comment]
Thanks @zerobfd 

I've tried that version of the script. Looks like it does grant "meshconfig.admin" to the service account.

Unfortunately it doesn't appear to have made a difference, because I still get the 401 Unauthorized:

```
install_asm: Initializing meshconfig API...
install_asm: Running: 'curl --request POST --fail --data  -o /dev/null https://meshconfig.googleapis.com/v1alpha1/projects/aivahealth-dev:initialize --header @-'
install_asm: -------------
curl: (22) The requested URL returned error: 401 Unauthorized
```

I'm wondering if it has anything to do with curl itself, like the bearer token isn't being passed in as a header properly or something.

**Comment by zhaytee** [comment]
Also, I'm doing this on Ubuntu 16.04 with curl 7.47.0

**Comment by zerobfd** [comment]
Sorry for the delayed response!

I can't come up with an easy reason why this might be happening. If you run the commands manually in verbose mode does it give any more insight?

```
gcloud auth activate-service-account \                                                                                 
    --project=myprojectid \                                                                                              
    anthos-installer@myprojectid.iam.gserviceaccount.com \                                                                                                   
    --key-file=./sa.json

curl --request POST -v \                                                                                                                                                                          
    "https://meshconfig.googleapis.com/v1alpha1/projects/myprojectid:initialize" \                                     
    --header @- <<EOF                                                                                                    
Authorization: Bearer $(gcloud --project=myprojectid auth print-access-token)                                                                                           
EOF                                                                                                                      
```

**Comment by zhaytee** [comment]
I played around a bit, and the issue has to do with the `--header @-` construct. When I remove that and replace it with the following, it works just fine:

```
curl --request POST --data '' -v \
    "https://meshconfig.googleapis.com/v1alpha1/projects/myprojectid:initialize" \
    --header "Authorization: Bearer $(gcloud --project=myprojectid auth print-access-token)"
```

(Also note that I had to add `--data ''` in there.)

So the problem is that the Authorization header just isn't getting passed into the request when the `--header @-` construct is used. 🧐 

I can manually patch the script on my end to make it work, although I'd obviously rather use the official unmodified script. What do you think the fix would be here? Am I using some janky old version of curl and/or bash?

Ubuntu 16.04 with curl 7.47.0 and bash 4.3.48

Thanks for your help!

**Comment by zerobfd** [comment]
(Oops, sorry about the --data '')

I didn't actually know this, but it turns out the @- construct was added in 7.55.

The entire reason I'm using it is to avoid accidentally printing the authorization token if someone runs the script with -v (since then it will be part of the command). But if curl < 7.55 is common then that's probably not worth breaking compatibility for.

Let me see if there's a cleverer way to hide that token and if not then I'll change it to specify the header directly. Thanks for finding this!

**Comment by zhaytee** [comment]
Sounds good, thanks @zerobfd 

FWIW, it does look like curl 7.47 is the latest version available in Ubuntu 16.04, which is an LTS release and thus is still probably pretty popular.

**Comment by zerobfd** [comment]
I just linked a PR to this issue (#301 ) that I think should fix it. If you have time to test the command once with your setup before I merge it that would be a huge help. I'm pretty sure that -K has been around for quite awhile, but I don't have an Ubuntu LTS setup available to test.

**Comment by zhaytee** [comment]
Thanks @zerobfd , this does seem to fix the 401 Unauthorized error!

Of course, now I'm encountering a different issue, but will open a new issue for that later. ;)
